crates/katana/rpc/rpc/src/torii.rs

```diff
@@ -0,0 +1,131 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use futures::StreamExt;
+use jsonrpsee::core::{async_trait, RpcResult};
+use katana_core::sequencer::KatanaSequencer;
+use katana_primitives::block::BlockHashOrNumber;
+use katana_provider::traits::block::BlockProvider;
+use katana_provider::traits::transaction::TransactionProvider;
+use katana_rpc_api::torii::ToriiApiServer;
+use katana_rpc_types::error::torii::ToriiApiError;
+use katana_rpc_types::transaction::{TransactionsPage, TransactionsPageCursor};
+use katana_tasks::TokioTaskSpawner;
+
+#[derive(Clone)]
+pub struct ToriiApi {
+    sequencer: Arc<KatanaSequencer>,
+}
+
+impl ToriiApi {
+    pub fn new(sequencer: Arc<KatanaSequencer>) -> Self {
+        Self { sequencer }
+    }
+
+    async fn on_io_blocking_task<F, T>(&self, func: F) -> T
+    where
+        F: FnOnce(Self) -> T + Send + 'static,
+        T: Send + 'static,
+    {
+        let this = self.clone();
+        TokioTaskSpawner::new().unwrap().spawn_blocking(move || func(this)).await.unwrap()
+    }
+}
+
+#[async_trait]
+impl ToriiApiServer for ToriiApi {
+    async fn get_transactions(
+        &self,
+        cursor: TransactionsPageCursor,
+    ) -> RpcResult<TransactionsPage> {
+        self.on_io_blocking_task(move |this| {
+            let mut transactions = Vec::new();
+            let mut next_cursor = cursor.clone();
+
+            let provider = this.sequencer.backend.blockchain.provider();
+            let latest_block_number = this.sequencer.block_number().map_err(ToriiApiError::from)?;
+
+            if cursor.block_number > latest_block_number + 1 {
+                return Err(ToriiApiError::BlockNotFound.into());
+            }
+
+            if latest_block_number >= cursor.block_number {
+                for block_number in cursor.block_number..=latest_block_number {
+                    let tx_range = BlockProvider::block_body_indices(
+                        provider,
+                        BlockHashOrNumber::Num(block_number),
+                    )
+                    .map_err(ToriiApiError::from)?
+                    .ok_or(ToriiApiError::BlockNotFound)?;
+
+                    let mut block_transactions = provider
+                        .transaction_in_range(Range::from(tx_range))
+                        .map_err(ToriiApiError::from)?
+                        .into_iter()
+                        .map(|tx| tx.clone().into())
+                        .collect::<Vec<_>>();
```

Comment:
```suggestion
                    let mut block_transactions = provider
                        .transactions_by_block(BlockHashOrNumber::Num(block_number));
```

you're only fetching transactions of a block and not from an arbitrary range. maybe this would be more convenient ?